card/mod/bootstrap/set/abstract/bootswatch_theme.rb

```
@@ -0,0 +1,167 @@
+def engine_input
+  extended_input_cards.select { |ca| ca.type_id.in? [ScssID, CssID] }
+end
+
+# Don't create "+*machine output" file card
+# instead save the the output as the card's content is
+def after_engine output
+  Card::Auth.as_bot { update_attributes! db_content: output }
+end
+
+# needed to make the refresh_script_and_style method work with these cards
+def source_files
+  extended_input_cards.map do |i_card|
+    i_card.try(:source_files)
+  end.flatten.compact
+end
+
+# needed to make the refresh_script_and_style method work with these cards
+def existing_source_paths
+  extended_input_cards.map do |i_card|
+    i_card.try(:existing_source_paths)
+  end.flatten.compact
+end
+
+def extended_input_cards
+  input_names.map do |n|
+    Card.fetch(n).extended_item_cards
+  end.flatten.compact
+end
+
+def content
+  CONTENT_PARTS.map do |n|
+    send "#{n}_content"
+  end.join "\n"
+end
+
+def pre_variables_content
+  load_content(*PRE_VARIABLES_CARD_NAMES)
+end
+
+def variables_content
+  load_content variables_card_names
+end
+
+def post_variables_content
+  load_content(*POST_VARIABLES_CARD_NAMES)
+end
+
+def stylesheets_content
+  load_content stylesheets_card_names
+end
+
+def input_names _args={}
+  (PRE_VARIABLES_CARD_NAMES + variables_card_names +
+    POST_VARIABLES_CARD_NAMES + stylesheets_card_names).compact.map do |n|
+    Card.fetch_name(n)
+  end.compact
+end
+
+def item_names _args={}
+  []
+end
+
+def item_cards _args={}
+  [self]
+end
+
+def load_content *names
+  cards = names.map { |n| Card.fetch(n)&.extended_item_cards }
+  cards.flatten.compact.map(&:content).join "\n"
+end
+
+def scss_from_theme_file file
+  return "" unless (path = ::File.join(source_dir, "_#{file}.scss")) &&
+                   ::File.exist?(path)
+  ::File.read path
+end
+
+def theme_name
+  /^(.+)_skin$/.match(codename)&.capture(0) ||
+    /^(.+)[ _][sS]kin/.match(name).capture(0)&.downcase
```

no ampersand after match?

---

...or I suppose that really should raise an error...